sdk/src/logs/batch_log_processor.cc

```diff
@@ -0,0 +1,238 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "opentelemetry/sdk/logs/batch_log_processor.h"
+
+#include <iostream>
+#include <vector>
+
+using opentelemetry::sdk::common::AtomicUniquePtr;
+using opentelemetry::sdk::common::CircularBufferRange;
+
+OPENTELEMETRY_BEGIN_NAMESPACE
+namespace sdk
+{
+namespace logs
+{
+/************************** Constructor *******************************/
+
+BatchLogProcessor::BatchLogProcessor(std::unique_ptr<LogExporter> &&exporter,
+                                     const size_t max_queue_size,
+                                     const std::chrono::milliseconds scheduled_delay_millis,
+                                     const size_t max_export_batch_size)
+    : exporter_(std::move(exporter)),
+      max_queue_size_(max_queue_size),
+      scheduled_delay_millis_(scheduled_delay_millis),
+      max_export_batch_size_(max_export_batch_size),
+      buffer_(max_queue_size_),
+      worker_thread_(&BatchLogProcessor::DoBackgroundWork, this)
+{}
+
+/************************** Helper methods used by the worker thread ***************/
+
+void BatchLogProcessor::Export(const bool was_force_flush_called)
+{
+  // Figure out the number of records to export
+  size_t num_records_to_export;
+
+  if (was_force_flush_called == true)
+  {
+    num_records_to_export = buffer_.size();
+  }
+  else
+  {
+    // Export the min of max_export_batch_size and buffer_size
+    num_records_to_export =
+        buffer_.size() >= max_export_batch_size_ ? max_export_batch_size_ : buffer_.size();
+  }
+
+  // Get records from the circular buffer and put into a vector
+  std::vector<std::unique_ptr<Recordable>> records_arr;
+
+  buffer_.Consume(
+      num_records_to_export, [&](CircularBufferRange<AtomicUniquePtr<Recordable>> range) noexcept {
+        range.ForEach([&](AtomicUniquePtr<Recordable> &ptr) {
+          std::unique_ptr<Recordable> swap_ptr = std::unique_ptr<Recordable>(nullptr);
+          ptr.Swap(swap_ptr);
+          records_arr.push_back(std::unique_ptr<Recordable>(swap_ptr.release()));
+          return true;
+        });
+      });
+
+  // Call exporter with the records, and get export status
+  if (exporter_->Export(nostd::span<std::unique_ptr<Recordable>>(
+          records_arr.data(), records_arr.size())) != ExportResult::kSuccess)
+  {
+    // Indicate Error: "[Batch Log Processor]: Failed to export a batch"
+  }
+
+  // In case this export was the result of a force flush, notify main thread that force flush is
+  // complete
+  if (was_force_flush_called == true)
+  {
+    is_force_flush_notified_.store(true);
+
+    while (is_force_flush_notified_.load() == true)
+    {
+      force_flush_cv_.notify_one();
```

Comment:
Why notify main thread by `force_flush_cv_.notify_one()` for multiple times, should just one notification work?

Reply:
I had the same question originally, but the `while` around it is to prevents the notification call being missed, in case the `notify` runs before the `wait`

Reply:
I think it is fine for the `notify_one()` be missed, because `force_flush_cv_.wait()` has checks `is_force_lufsh_notified` is its predicate, which is already set before `notify_one()`. So if `notify_one()` happened before the `force_flush_cv_wait()`, the latter wait will return immediately without entering wait state.

Reply:
I removed it in [this commit](https://github.com/open-telemetry/opentelemetry-cpp/pull/434/commits/14033b514096017b57d9c8cd00217e383b2798e4#diff-f7c189661e15834743b64c453ad9e3548337c19515074da4bc43bb0205c94ffaR85) and it seemed to be working. But following the [batch span processor](https://github.com/open-telemetry/opentelemetry-cpp/blob/master/sdk/src/trace/batch_span_processor.cc#L163), which wraps all the `cv.notify()` with `while` loops`, I thought it should be kept consistent

Reply:
https://github.com/open-telemetry/opentelemetry-cpp/issues/477